tools/user_migration/migrate_user.py

```diff
@@ -0,0 +1,374 @@
+#! venv/bin/python
+
+# painfully stupid script to check out dumping mysql databases to avro.
+# Avro is basically "JSON" for databases. It's not super complicated & it has
+# issues (one of which is that it requires Python2).
+#
+#
+
+import argparse
+import logging
+import base64
+import sys
+import time
+from datetime import datetime
+
+from mysql import connector
+from google.cloud import spanner
+from google.api_core.exceptions import AlreadyExists
+from urllib.parse import urlparse
+
+SPANNER_NODE_ID = 800
+
+
+class BadDSNException(Exception):
+    pass
+
+
+class Collections:
+    """Cache spanner collection list.
+
+    The spanner collection list is the (soon to be) single source of
+    truth regarding collection ids.
+
+    """
+    _by_name = {}
+    databases = None
+
+    def __init__(self, databases):
+        """Get the cache list of collection ids"""
+        sql = """
+        SELECT
+            name, collection_id
+        FROM
+            collections;
+        """
+        self.databases = databases
+        logging.debug("Fetching collections...")
+        with self.databases['spanner'].snapshot() as cursor:
+            rows = cursor.execute_sql(sql)
+            for row in rows:
+                self._by_name[row[0]] = row[1]
+
+    def get_id(self, name):
+        """ Get/Init the ID for a given collection """
+        if name in self._by_name:
+            return self._by_name.get(name)
+
+        # option 1: do what spanner-rs does
+        with self.databases['spanner'].batch() as cursor:
```

Comment:
Does batch() support execute_sql? You may need to switch to transaction() instead

Reply:
Batch does support `execute_sql`. I believe it's part of the base class for `transaction`, `snapshot` and `batch`.

Reply:
Batch is different, _BatchBase only supports mutation operations: https://github.com/googleapis/google-cloud-python/blob/ee6214251fbe83cb83a9b5b5019c605a27a01354/spanner/google/cloud/spanner_v1/batch.py#L29 (_SessionWrapper has nothing)